File: check_bipartite.go
package main

import (
	"fmt"
)

func checkBipartite(graph [][]int) bool {
	colorCodes := make(map[int]int) // 1 for red, -1 for blue
	stack := []int{}
	for i := 0; i < len(graph); i++ {
		stack = append(stack, i)
	}

	for len(stack) > 0 {
		node := stack[len(stack)-1]
		stack = stack[:len(stack)-1]

		if _, ok := colorCodes[node]; !ok {
			colorCodes[node] = 1
		}

		for _, neighbor := range graph[node] {
			if _, ok := colorCodes[neighbor]; ok {
				if colorCodes[node] == colorCodes[neighbor] {
					return false
				}
			} else {
				colorCodes[neighbor] = -colorCodes[node]
				stack = append(stack, neighbor)
			}
		}
	}

	return true
}

func main() {
	//{{1},{0,3},{3},{1,2}}
	// graph := [][]int{{1, 3}, {0, 2}, {1, 3}, {0, 2}}
	graph := [][]int{{1}, {0, 3}, {3}, {1, 2}}

	// graph := [][]int{{1, 2, 3}, {0, 2}, {0, 1, 3}, {0, 2}}
	isBipartite := checkBipartite(graph)
	fmt.Printf("%v", isBipartite)
}

File: check_bipartite.js
function isBipartite(graph) {
  const colorCodes = {} // 1 for red, -1 for blue

  const queue = []
  for (let i = graph.length - 1; i >= 0; i--) {
    queue.push(parseInt(i))
  }

  while (queue.length) {
    const node = queue.pop()
    colorCodes[node] = colorCodes[node] || 1
    for (const neighbor of graph[node]) {
      if (!colorCodes[neighbor]) {
        colorCodes[neighbor] = -colorCodes[node]
        queue.push(neighbor)
      } else if (colorCodes[neighbor] === colorCodes[node]) {
        return false
      }
    }
  }

  return true
}

graph = [
  [1, 2, 3],
  [0, 2],
  [0, 1, 3],
  [0, 2],
]

// graph = [
//   [1, 3],
//   [0, 2],
//   [1, 3],
//   [0, 2],
// ]

// graph = [[1], [0, 3], [3], [1, 2]]

res = isBipartite(graph)
console.log(res)
